File: embedded/renderer/factories/ElementNodeFactory.cpp
#include "ElementNodeFactory.hpp"

#include <string>

#include "renderer/nodes/BjorklundRenderNode.hpp"
#include "renderer/nodes/ElementRenderNode.hpp"
#include "renderer/nodes/StretchRenderNode.hpp"

namespace krill
{
  namespace factory
  {
    namespace
    {
      Fraction sourceUnitsForFixedStep(
        const rapidjson::Value& source,
        std::optional<Fraction> (*valueAsFraction)(const rapidjson::Value&))
      {
        if (!source.IsObject())
        {
          return Fraction(1);
        }

        if (source.HasMember("type_") && source["type_"].IsString())
        {
          const auto type = std::string(source["type_"].GetString());

          if (type == "element" && source.HasMember("source_"))
          {
            return sourceUnitsForFixedStep(source["source_"], valueAsFraction);
          }

          if (type == "pattern"
              && source.HasMember("arguments_")
              && source["arguments_"].IsObject()
              && source["arguments_"].HasMember("alignment")
              && source["arguments_"]["alignment"].IsString()
              && std::string(source["arguments_"]["alignment"].GetString()) == "h"
              && source.HasMember("source_")
              && source["source_"].IsArray())
          {
            Fraction total(0);
            for (const auto& child : source["source_"].GetArray())
            {
              Fraction weight(1);
              if (child.IsObject() && child.HasMember("options_") && child["options_"].IsObject())
              {
                const auto& options = child["options_"];
                if (options.HasMember("weight"))
                {
                  const auto maybeWeight = valueAsFraction(options["weight"]);
                  if (maybeWeight.has_value() && maybeWeight.value() > Fraction(0))
                  {
                    weight = maybeWeight.value();
                  }
                }
              }
              total += weight;
            }
            if (total > Fraction(0))
            {
              return total;
            }
          }
        }

        return Fraction(1);
      }

      RenderNodePtr applyElementOperator(
        RenderNodePtr node,
        const rapidjson::Value& elementNode,
        std::optional<Fraction> (*valueAsFraction)(const rapidjson::Value&),
        std::optional<long> (*valueAsLong)(const rapidjson::Value&))
      {
        if (!elementNode.IsObject() || !elementNode.HasMember("options_") || !elementNode["options_"].IsObject())
        {
          return node;
        }

        const auto& options = elementNode["options_"];
        if (!options.HasMember("operator") || !options["operator"].IsObject())
        {
          return node;
        }

        const auto& op = options["operator"];
        if (!op.HasMember("type_") || !op["type_"].IsString())
        {
          return node;
        }
        if (!op.HasMember("arguments_") || !op["arguments_"].IsArray())
        {
          return node;
        }

        const auto opType = std::string(op["type_"].GetString());
        const auto& args = op["arguments_"].GetArray();

        if (opType == "bjorklund" && args.Size() >= 2)
        {
          const auto pulses = valueAsLong(args[0]);
          const auto steps = valueAsLong(args[1]);
          if (pulses.has_value() && steps.has_value())
          {
            return std::make_shared<BjorklundRenderNode>(node, pulses.value(), steps.value());
          }
        }

        if (opType == "stretch" && args.Size() > 0)
        {
          const auto factor = valueAsFraction(args[0]);
          if (factor.has_value())
          {
            return std::make_shared<StretchRenderNode>(node, factor.value());
          }
        }

        if (opType == "fixed-step" && args.Size() > 0)
        {
          const auto stepSize = valueAsFraction(args[0]);
          if (stepSize.has_value() && stepSize.value() > Fraction(0))
          {
            const auto units = sourceUnitsForFixedStep(elementNode["source_"], valueAsFraction);
            return std::make_shared<StretchRenderNode>(node, units / stepSize.value());
          }
        }

        return node;
      }
    }

    std::optional<RenderNodePtr> makeElementNode(
      const rapidjson::Value& v,
      RenderNodePtr (*buildRenderTree)(const rapidjson::Value&),
      std::string (*sourceAsString)(const rapidjson::Value&),
      std::optional<Fraction> (*valueAsFraction)(const rapidjson::Value&),
      std::optional<long> (*valueAsLong)(const rapidjson::Value&))
    {
      if (!v.IsObject())
      {
        return std::nullopt;
      }
      if (!v.HasMember("type_") || !v["type_"].IsString())
      {
        return std::nullopt;
      }
      if (std::string(v["type_"].GetString()) != "element")
      {
        return std::nullopt;
      }
      if (!v.HasMember("source_"))
      {
        return std::nullopt;
      }

      const auto& source = v["source_"];
      if (source.IsObject())
      {
        auto element = std::make_shared<ElementRenderNode>(buildRenderTree(source));
        return applyElementOperator(element, v, valueAsFraction, valueAsLong);
      }

      auto element = std::make_shared<ElementRenderNode>(sourceAsString(source));
      return applyElementOperator(element, v, valueAsFraction, valueAsLong);
    }
  }
}

File: core/js/renderer/factories/element-node-factory.js
var math = require("mathjs");
var factoryUtils = require("./factory-utils.js");
var NoteVelocity = require("../../utils/note-velocity.js");
var TypeGuards = require("../../utils/type-guards.js");

require("../nodes/bjorklund-render-node.js");
require("../nodes/element-render-node.js");
require("../nodes/stretch-render-node.js");

function elementControlsFromModel(modelNode, isNestedSource)
{
  if (!modelNode || !TypeGuards.isPlainObject(modelNode.controls_))
  {
    return null;
  }

  var controls = Object.assign({}, modelNode.controls_);

  // Language-level convenience: group velocity is interpreted as a factor.
  if (isNestedSource && controls.velocity !== undefined)
  {
    var nestedVelocityFactor = NoteVelocity.resolveVelocityFactor(controls.velocity);
    controls.velocityFactor = controls.velocityFactor !== undefined
      ? NoteVelocity.accumulateVelocityFactor(controls.velocityFactor, nestedVelocityFactor)
      : nestedVelocityFactor;
    delete controls.velocity;
  }

  if (isNestedSource && controls.velocityFactor !== undefined)
  {
    controls.velocityFactor = NoteVelocity.resolveVelocityFactor(controls.velocityFactor);
  }

  return Object.keys(controls).length > 0 ? controls : null;
}

function sourceUnitsForFixedStep(modelNode)
{
  if (!factoryUtils.isObjectNode(modelNode))
  {
    return math.fraction(1);
  }

  if (modelNode.type_ === "element" && TypeGuards.isPlainObject(modelNode.source_))
  {
    return sourceUnitsForFixedStep(modelNode.source_);
  }

  if (modelNode.type_ === "pattern"
      && modelNode.arguments_
      && modelNode.arguments_.alignment === "h"
      && Array.isArray(modelNode.source_)
      && modelNode.source_.length > 0)
  {
    var total = math.fraction(0);
    modelNode.source_.forEach(function(child)
    {
      var weight = math.fraction(1);
      if (child
          && factoryUtils.isObjectNode(child)
          && child.options_
          && child.options_.weight !== undefined)
      {
        try
        {
          weight = math.fraction(child.options_.weight);
        }
        catch (err)
        {
          weight = math.fraction(1);
        }
        if (math.smallerEq(weight, 0))
        {
          weight = math.fraction(1);
        }
      }
      total = math.add(total, weight);
    });
    if (math.larger(total, 0))
    {
      return total;
    }
  }

  return math.fraction(1);
}

function applyElementOperator(node, modelNode)
{
  if (!modelNode || !modelNode.options_ || !modelNode.options_.operator)
  {
    return node;
  }

  var operator = modelNode.options_.operator;
  if (!factoryUtils.isObjectNode(operator) || !operator.type_)
  {
    return node;
  }

  if (operator.type_ === "bjorklund" && Array.isArray(operator.arguments_) && operator.arguments_.length >= 2)
  {
    return new BjorklundRenderNode(node, operator.arguments_[0], operator.arguments_[1]);
  }

  if (operator.type_ === "stretch" && Array.isArray(operator.arguments_) && operator.arguments_.length > 0)
  {
    return new StretchRenderNode(node, operator.arguments_[0]);
  }

  if (operator.type_ === "fixed-step" && Array.isArray(operator.arguments_) && operator.arguments_.length > 0)
  {
    var stepSize;
    try
    {
      stepSize = math.fraction(operator.arguments_[0]);
    }
    catch (err)
    {
      return node;
    }
    if (math.smallerEq(stepSize, 0))
    {
      return node;
    }
    var units = sourceUnitsForFixedStep(modelNode.source_);
    return new StretchRenderNode(node, math.divide(units, stepSize));
  }

  return node;
}

function makeElementNode(modelNode, buildRenderNode)
{
  if (!factoryUtils.hasType(modelNode, "element"))
  {
    return null;
  }

  var elementNode;
  var isNestedSource = TypeGuards.isPlainObject(modelNode.source_);
  var controls = elementControlsFromModel(modelNode, isNestedSource);
  if (isNestedSource)
  {
    elementNode = new ElementRenderNode(buildRenderNode(modelNode.source_), controls);
    return applyElementOperator(elementNode, modelNode);
  }

  elementNode = new ElementRenderNode(modelNode.source_, controls);
  return applyElementOperator(elementNode, modelNode);
}

module.exports = {
  makeElementNode: makeElementNode
};
